src/mc-tokens.c

```diff
@@ -0,0 +1,312 @@
+/*
+ * Copyright 2022-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mc-tokens-private.h"
+#include "mongocrypt-endian-private.h"
+
+/// Define a token type of the given name, with constructor parameters given as
+/// the remaining arguments. This macro usage should be followed by the
+/// constructor body, with the implicit first argument '_mongocrypt_crypto_t*
+/// crypto' and final argument 'mongocrypt_status_t* status'
+#define DEF_TOKEN_TYPE(Name, ...) \
+   DEF_TOKEN_TYPE_1 (Name, CONCAT (Name, _t), __VA_ARGS__)
+
+#define DEF_TOKEN_TYPE_1(Prefix, T, ...)                             \
+   /* Define the struct for the token */                             \
+   struct T {                                                        \
+      _mongocrypt_buffer_t data;                                     \
+   };                                                                \
+   /* Data-getter */                                                 \
+   const _mongocrypt_buffer_t *CONCAT (Prefix, _get) (const T *self) \
+   {                                                                 \
+      return &self->data;                                            \
+   }                                                                 \
+   /* Destructor */                                                  \
+   void CONCAT (Prefix, _destroy) (T * self)                         \
+   {                                                                 \
+      if (!self) {                                                   \
+         return;                                                     \
+      }                                                              \
+      _mongocrypt_buffer_cleanup (&self->data);                      \
+      bson_free (self);                                              \
+   }                                                                 \
+   /* Constructor. Parameter list given as variadic args. */         \
+   T *CONCAT (Prefix, _new) (_mongocrypt_crypto_t * crypto,          \
+                             __VA_ARGS__,                            \
+                             mongocrypt_status_t * status)
+
+DEF_TOKEN_TYPE (mc_CollectionsLevel1Token, const _mongocrypt_buffer_t *RootKey)
+{
```

Comment:
I know I snickered at the macro stacking in slack a little, but I think we can probably go a little further to mirror what we have in the lovely comment above.  Take this as a shared idea, not a requirement for refactor.  It's got pros-and-cons to it.

```
#define IMPL_TOKEN_NEW_1 (Name, Key, Arg, Clean) \
{ \
  CONCAT (Name, _t)* t = bson_malloc (sizeof (CONCAT (Name, T))); \
  _mongocrypt_buffer_init (&t->data); \
  _mongocrypt_buffer_resize (&t->data, MONGOCRYPT_HMAC_SHA256_LEN); \
\
   if (!_mongocrypt_hmac_sha_256 ( \
          crypto, Key, Arg, &t->data, status)) { \
      CONCAT(Name, _destroy) (t); \
      Clean;
      return NULL; \
   } \
   Clean;
   return t; \
}

// Use this form for hashing with a mongocrypt buffer with dynamic contents.
#define IMPL_TOKEN_NEW_1 (Name, Key, Arg) \
    IMPL_TOKEN_NEW_1 (name, Key, Arg)
    
// Use this form for lower level key derivation using fixed constants.
#define IMPL_TOKEN_NEW_CONST(Name, Key, Arg) \
{ \
    _mongocrypt_buffer_t to_hash; \
    _mongocrypt_buffer_copy_from_uint64 (&to_hash, MONGOCRYPT_UINT64_TO_LE (Arg) ); 
    IMPL_TOKEN_NEW_1 (Name, Key, to_hash, _mongocrypt_buffer_cleanup (&to_hash)) \
}
```

Then each type implementation can be:

```
DEF_TOKEN_TYPE (mc_CollectionsLevel1Token, const _mongocrypt_buffer_t *RootKey)
IMPL_TOKEN_NEW_CONST (mc_CollectionsLevel1Token, RootKey, 1)

DEF_TOKEN_TYPE (mc_ServerDataEncryptionLevel1Token, const _mongocrypt_buffer_t *RootKey)
IMPL_TOKEN_NEW_CONST (mc_ServerDataEncryptionLevel1Token, RootKey, 3)

DEF_TOKEN_TYPE (mc_EDCToken, const mc_CollectionsLevel1Token_t *CollectionsLevel1Token)
IMPL_TOKEN_NEW_CONST (mc_EDCToken, mc_CollectionsLevel1Token_get (CollectionsLevel1Token), 1)

// etc...

DEF_TOKEN_TYPE (mc_EDCDerivedFromDataToken, const mc_EDCToken_t *EDCToken, const _mongocrypt_buffer_t *v)
IMPL_TOKEN_NEW (mc_EDCDerivedFromDataToken, mc_EDCToken_get(EDCToken), v)

// etc...
```

Plus or minus a bit to improve readability and bring the usages closer to looking like their documented formats.

Reply:
Thank you for the suggested code. I like the idea of getting this closer to the documented formats. Updated.